README.md

```diff
@@ -8,50 +8,67 @@ You must install the Emacs plugin (this) and a language server
 
 ### lsp-scala
 
-Install the `lsp-mode` and `sbt-mode` dependencies using your preferred package
-manager. Then clone this repo and load it:
+Clone this repo, and fetch `lsp-mode`, `lsp-ui`, and `sbt-mode` from
+MELPA with the package manager of your choice.  (We are working on
+adding this package to MELPA to make `lsp-scala` load like the rest.)
+Here is an example using `use-package`:
 
 ```emacs-lisp
-(add-to-list 'load-path "<path to lsp-scala>")
-(require 'lsp-scala)
+(use-package lsp-mode
+  :ensure t)
+
+(use-package lsp-scala
+  :load-path "~/path/to/lsp-scala"
+  :config
+  (setq lsp-scala-server-command "~/bin/metals-emacs")
+  (require 'lsp)
+  (add-hook 'scala-mode #'lsp))
```

Comment:
I tried to use `:hook` here, but that caused this package to be deferred, and then we don't register to `lsp-clients` in time.  Maybe @JesusMtnez knows a trick.

Reply:
I guess the problem is that it's registered from `eval-after-load`. You can register a client without `eval-after-load`. It's always possible to add `:after lsp`, for (I think) the same behaviour.

Please note that I'm nowhere near elisp expert and I don't know when lsp-mode expects clients to be registered 😸

Reply:
I struggled with this for a while.

* I agree that `:after lsp` should work, but it doesn't.
* `lsp` doesn't autoload, and I'm not sure why.  I can put `(require 'lsp)` in either the `use-package lsp-mode` or the `use-package lsp-scala`, but I couldn't get it working without it somewhere.
* Even with the explicit require, `lsp-clients` would not be populated without doing the `eval-after-load`.  I don't understand why.
* I thought `(when (require 'lsp nil t) ...)` might work in lsp-scala.el.  It does not.
* Maybe it would work with just a straight `(require 'lsp)` instead of `(require 'lsp-mode)` in lsp-scala.el.  This would break compatibility with MELPA stable, but I'm not sure that we care as long as we update the metals docs promptly.
* Maybe this is all complicated by not being able to `:ensure t` because we're not in MELPA yet?

I don't claim to know what I'm doing. 😆

Reply:
I'm not sure if it's better, but this will also work:
```elisp
(use-package lsp-scala
  :hook (scala-mode . (lambda ()
                        (require 'lsp-scala)
                        (lsp)))
  :load-path "~/lsp-scala"
  :config
  (setq lsp-scala-server-command "~/bin/metals-emacs"))
```

Reply:
I was able to make it work using @wedens snippet, applying `(require 'lsp-scala)` before invoking `(lsp)`, so the client is registered.
My current config is like:

```el
(use-package lsp-mode)

(use-package lsp-ui
  :hook (lsp-mode . lsp-ui-mode))

(use-package lsp-scala
  :load-path "local/lsp-scala/"
  :hook (scala-mode . (lambda()
                        (require 'lsp-scala)
                        (lsp))))
```

It seems it works, but I'm getting `lsp--*` errors (not related to `lsp-scala` init, for sure. They seem to be about client-server communication). I'll continue trying to improve that `lambda`.

Reply:
The thing is that `(require 'lsp-scala)` has to be done somewhere. Doing it in `lsp-mode-hook` seems to be too late. I can't find a better place for it that doesn't compromise deferred loading.

Reply:
I was able to register the client properly using:

```el
(use-package lsp-scala
  :load-path "local/lsp-scala"
  :after lsp ; load after `lsp` is loaded
  :demand t ; force `(require 'lsp-scala)` after `lsp` is loaded
  :hook (scala-mode . #'lsp)) ; not working :worried: 
```

This way, `lsp-scala` is loaded after `lsp` is loaded (so the functions are available). Using `:demand t` is necessary since there is no command or bind that will trigger the package to load. Using it, will make the `(require 'lsp-scala` happen after `lsp` is loaded.

Now I'm working on getting the hook to trigger the function `lsp` so it automatically starts working. Right now, need to `M-x lsp` to get metals running.

Reply:
I guess you'll need to move hook to `:init`.

Reply:
Got it working! :+1: 

The problem of using `:after lsp` was that no one is actually loading `lsp`, since `scala-mode` does not require it at all. Instead of depending on `lsp`, `lsp-scala` should load after `scala-mode`. It loads everything with the following snippet:

```el
(use-package lsp-scala
  :after scala-mode
  :demand t
  :load-path "local/lsp-scala/"
  :hook (scala-mode . lsp))
```